File: backend/rubik/RubikMoves.py
from parser import check_notation, check_basic_notation, check_parenthesis_notation
from ErrorManager import NotationError
import re

class RubikMoves:
	"""The class to init a rubik sequence
	"""
	def __init__(self, notation: str) -> None:
		self.notation: str = notation
		self.repeats: int = 1
		self.sequences: list[str] = []
		if check_notation(notation) == False:
			raise NotationError(f"{notation} is not a valid notation.")

		self.extract_repeats()
		self.extract_sequences()


	def extract_repeats(self) -> int:
		match = re.search(r"\d+\'?$", self.notation)
		if match:
			repeats = re.sub(r'\'', '', match.group())
			self.repeats = int(repeats)
		# 	print(self.repeats)
		# else:
		# 	print("No repeat.")

		return self.repeats

	def __get_move(self, move) -> list[str]:
		match_nb = re.search(r"\d+", move)
		repeats = 1
		if match_nb:
			repeats = re.sub(r'\'', '', match_nb.group())
			repeats = int(repeats)

		clear_move = re.sub(r"\d", '', move)
		# print(f"{move} | {clear_move} : {repeats}")
		return [clear_move for i in range(repeats)]

	def extract_sequences(self) -> list[str]:
		self.sequences.clear()
		regex = r"[FRUBLDfrubld]\d*\'?"
		moves_match = re.findall(regex, self.notation)
		if moves_match:
			print(moves_match)
			for move in moves_match:
				self.sequences += self.__get_move(move)
		self.sequences *= self.repeats
		return self.sequences

	def __str__(self) -> str:
		return self.notation

	def __repr__(self) -> str:
		return self.__str__()


if __name__ == "__main__":
	RubikMoves("r4'")
	RubikMoves("r'")
	RubikMoves("(r')")
	RubikMoves("(RU42R'U'R41')4")
	pass

File: frontend/app/src/js/RubikMoves.ts
function checkBasicNotation(notation: string) {
    // Check basic notation
    const regex = /^[FRUBLD]\d*'?$/;
    return regex.test(notation);
}

function checkParenthesisNotation(notation: string) {
    // Check with parenthesis
    const regexParenthesis = /^\(([FRUBLD]\d*'?)+\)\d*$/;
    return regexParenthesis.test(notation);
}

function checkNotation(notation: string) {
    /**
     * Function to check if a notation is valid.
     *
     * Args:
     *   notation (string): Notation to check.
     *
     * Returns:
     *   boolean: True if the notation is valid.
     */
    return checkBasicNotation(notation) || checkParenthesisNotation(notation);
}


class RubikMoves {
    /**
     * The class to init a rubik sequence
     * @param {string} notation
     */

	notation: string;
	repeats: number;
	sequences: Array<string>;
    constructor(notation: string) {
        this.notation = notation;
        this.repeats = 1;
        this.sequences = [];

        if (!checkNotation(notation)) {
            throw new Error(`${notation} is not a valid notation.`);
        }

        this.extractRepeats();
        this.extractSequences();
    }

    extractRepeats() {
        const match = this.notation.match(/\d+'?$/);
        if (match) {
            const repeats = match[0].replace(/'/g, '');
            this.repeats = parseInt(repeats, 10);
        }
        return this.repeats;
    }

    private getMove(move: string) {
        const matchNb = move.match(/\d+/);
        let repeats = 1;
        if (matchNb) {
            repeats = parseInt(matchNb[0].replace(/'/g, ''), 10);
        }

        const clearMove = move.replace(/\d/, '');
        return Array(repeats).fill(clearMove);
    }

    extractSequences() {
        this.sequences = [];
        const regex = /[FRUBLDfrubld]\d*'?/g;
        const movesMatch = this.notation.match(regex);

        if (movesMatch) {
            for (const move of movesMatch) {
                this.sequences = this.sequences.concat(this.getMove(move));
            }
        }

		// console.log("OLD SEQ : " + this.sequences);

		// console.log(this.repeats);
		const tmp_sequences = this.sequences;
		if (this.notation.match(/\)\d+$/)) {
			for (let i = 1; i < this.repeats; i++) {
				this.sequences = this.sequences.concat(tmp_sequences);
			}
		}
		// console.log("NEW SEQ : " + this.sequences);
        return this.sequences;
    }

    toString() {
        return this.notation;
    }

    toJSON() {
        return this.toString();
    }
}

export { RubikMoves, checkNotation }
